src/crimson/os/seastore/journal/circular_bounded_journal.cc

```diff
@@ -103,85 +114,69 @@ CircularBoundedJournal::close_ertr::future<> CircularBoundedJournal::close()
+    return record_submitter.wait_available(
+    ).safe_then([this, record=std::move(record), &handle]() mutable {
+      return do_submit_record(std::move(record), handle);
+    });
+  }
+  auto action = record_submitter.check_action(record.size);
+  if (action == RecordSubmitter::action_t::ROLL) {
+    std::ignore = cjs.roll();
+    auto expected_size = record_group_size_t(
+      record.size,
+      get_block_size()
+    ).get_encoded_length();
+    if (expected_size > get_records_available_size()) {
       ERROR("rolled, record size {}, but available size {}",
-            encoded_size, get_records_available_size());
+	    expected_size, get_records_available_size());
       return crimson::ct_error::erange::make();
     }
   }
 
-  journal_seq_t j_seq = get_written_to();
-  ceph::bufferlist to_write = encode_record(
-    std::move(record), device->get_block_size(),
-    j_seq, 0);
-  assert(to_write.length() == encoded_size);
-  auto target = get_rbm_addr(get_written_to());
-  auto new_written_to = target + encoded_size;
-  if (new_written_to >= get_journal_end()) {
-    assert(new_written_to == get_journal_end());
-    DEBUG("roll");
-    paddr_t paddr = convert_abs_addr_to_paddr(
-      get_records_start(),
-      get_device_id());
-    set_written_to(
-      journal_seq_t{++written_to.segment_seq, paddr});
-  } else {
-    paddr_t paddr = convert_abs_addr_to_paddr(
-      new_written_to,
-      get_device_id());
-    set_written_to(
-      journal_seq_t{written_to.segment_seq, paddr});
-  }
-  DEBUG("{}, target {}", r_size, target);
-
-  auto write_result = write_result_t{
-    j_seq,
-    encoded_size
-  };
-  auto write_fut = device_write_bl(target, to_write);
+  DEBUG("H{} submit {} ...",
+	(void*)&handle,
+	action == RecordSubmitter::action_t::SUBMIT_FULL ?
+	"FULL" : "NOT_FULL");
+  auto submit_fut = record_submitter.submit(std::move(record));
   return handle.enter(write_pipeline->device_submission
-  ).then([write_fut = std::move(write_fut)]() mutable {
-    return std::move(write_fut);
-  }).safe_then([this, &handle] {
-    return handle.enter(write_pipeline->finalize);
-  }).safe_then([this, target,
-    length=encoded_size,
-    write_result,
-    r_size,
-    FNAME] {
-    DEBUG("commit target {} used_size {} written length {}",
-          target, get_records_used_size(), length);
-
-    paddr_t paddr = convert_abs_addr_to_paddr(
-      target + r_size.get_mdlength(),
-      get_device_id());
-    auto submit_result = record_locator_t{
-      paddr,
-      write_result
-    };
-    trimmer.set_journal_head(write_result.start_seq);
-    return submit_result;
+  ).then([submit_fut=std::move(submit_fut)]() mutable {
+    return std::move(submit_fut);
+  }).safe_then([FNAME, this, &handle](record_locator_t result) {
+    return handle.enter(write_pipeline->finalize
+    ).then([FNAME, result, &handle] {
+      DEBUG("H{} finish with {}", (void*)&handle, result);
+      return result;
+    });
   });
 }
```

Comment:
Looks do_submit_record() and submit_record() can be finally identical to SegmentedJournal.

Before we can really reuse these code, let's organize the code text as consistent as possible for easier cross-checking and mark out the parts that cannot be unified yet (probably with TODO).